clippy_lints/src/return_self_not_must_use.rs

```diff
@@ -0,0 +1,101 @@
+use clippy_utils::{diagnostics::span_lint, must_use_attr, nth_arg, return_ty};
+use rustc_hir::def_id::LocalDefId;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, FnDecl, HirId, TraitItem, TraitItemKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// This lint warns when a method returning `Self` doesn't have the `#[must_use]` attribute.
```

Comment:
> They look like different use cases to me. What I'm trying to prevent here is cases like 0f32.rad(2.). It's much less likely (in my opinion) if it's not a method.
> ~ `@GuillaumeGomez`
>
> Ah sure, if you don't want to cover that case here that makes sense. I also think it's quite unlikely that someone calls such a constructor function and forgets to actually do something with the result. Adding #[must_use] on all these might add quite a bit of noise to the code.
> ~`@sdroege`

(I copied parts of [the discussion](https://github.com/rust-lang/rust-clippy/pull/8071#discussion_r762047475), as I was unable to comment on it for some reason :thinking: )

I understand the reasoning behind this decision, and I'm also in favor of this solution. Could you mention this in the lint description as well, to avoid false negative reports? You can add a *Limitations* section before the example or mention here that only methods will be linted which take a `self` argument.

Reply:
Done!